web3-functions/gelato-cctp/index.ts

```diff
@@ -1,153 +1,197 @@
+  const currentBlock = await provider.getBlockNumber();
+  const lastBlockStr = await context.storage.get(chainId);
+  const lastBlock = lastBlockStr ? Number(lastBlockStr) : currentBlock;
+
+  if (!lastBlockStr)
+    await context.storage.set(chainId, currentBlock.toString());
+
+  // if no blocks have passed since last execution, return early
+  // no reason to check attestations since the attestation service waits for new blocks
+  if (currentBlock === lastBlock)
+    return { canExec: false, message: "No blocks to index" };
 
+  // get existing pending transfers and relay tasks
   const pendingStr = await context.storage.get("pending");
-  const pending: ITransfer[] = pendingStr ? JSON.parse(pendingStr) : [];
+  let pending: ITransfer[] = pendingStr ? JSON.parse(pendingStr) : [];
 
+  const taskIdsStr = await context.storage.get("tasks");
+  let taskIds: string[] = taskIdsStr ? JSON.parse(taskIdsStr) : [];
+
+  // instantiate all contracts on the current network
   const runner = { provider: provider as any };
 
-  const messageTransmitter = IMessageTransmitter__factory.connect(
-    constants.messageTransmitter,
+  const circleMessageTransmitter = IMessageTransmitter__factory.connect(
+    network.circleMessageTransmitter,
     runner
   );
 
-  const gelatoReceiver = GelatoCCTPReceiver__factory.connect(
-    constants.gelatoReceiver,
+  const gelatoCCTPReceiver = GelatoCCTPReceiver__factory.connect(
+    network.gelatoCCTPReceiver,
     runner
   );
 
-  const gelatoSender = GelatoCCTPSender__factory.connect(
-    constants.gelatoSender,
+  const gelatoCCTPSender = GelatoCCTPSender__factory.connect(
+    network.gelatoCCTPSender,
     runner
   );
 
-  // index all logs since last processed block
-  // both MessageTransmitter and GelatoCCTPSender emit on depositForBurn
-  // every GelatoCCTPSender event corresponds to a MessageTransmitter event
-  // but not every MessageTransmitter event corresponds to a GelatoCCTPSender
-  // we must merge these events together based on their transactionHash
-  const messages = await messageTransmitter.queryFilter(
-    messageTransmitter.filters.MessageSent,
-    fromBlock,
-    toBlock
+  // check each task status
+  const tasks = await Promise.all(
+    taskIds.map((taskId) => getRelayTaskStatus(taskId))
   );
 
-  const deposits = await gelatoSender.queryFilter(
-    gelatoSender.filters.DepositForBurn,
-    fromBlock,
-    toBlock
+  // keep pending tasks and remove executed or failed tasks
+  // signal on failure (this could trigger an API call to notify the user)
+  taskIds = taskIds.filter((taskId, i) => {
+    const task = tasks[i];
+    if (!task) return true;
+
+    if (
+      task.taskState === TaskState.CheckPending ||
+      task.taskState === TaskState.ExecPending ||
+      task.taskState === TaskState.WaitingForConfirmation
+    )
+      return true;
+
+    if (task.taskState !== TaskState.ExecSuccess)
+      console.error("Task failed:", taskId);
+
```

Comment:
Always try to avoid NOT so rewrite to: if success, console.log, else: console.error